src/lib.rs

```diff
@@ -182,65 +182,73 @@ impl  Device for Cp2130 {
-        let out = buff.to_vec();
-        let _n = self.inner.lock().unwrap().spi_write_read(&out, buff)?;
+    fn transfer_in_place<'w>(&mut self, data: &'w mut [u8]) -> Result<(), Self::Error> {
+        let mut read_to: Vec<u8> = Vec::with_capacity(data.len());
+        let read_to = read_to.as_mut_slice();
+        self.bus.lock().unwrap().spi_write_read(data, read_to)?;
+        data.copy_from_slice(read_to);
         Ok(())
     }
 }
 
-
-impl embedded_hal::spi::blocking::Write<u8> for Spi {
-
-    fn write(&mut self, words: &[u8] ) -> Result<(), Self::Error> {
-        let _n = self.inner.lock().unwrap().spi_write(words)?;
+impl embedded_hal::spi::blocking::SpiBusWrite<u8> for SpiProxy {
+    fn write(&mut self, words: &[u8]) -> Result<(), Self::Error> {
+        let _n = self.bus.lock().unwrap().spi_write(words)?;
         Ok(())
     }
 }
 
-impl embedded_hal::spi::blocking::Read<u8> for Spi {
-
-    fn read(&mut self, buff: &mut [u8] ) -> Result<(), Self::Error> {
+impl embedded_hal::spi::blocking::SpiBusRead<u8> for SpiProxy {
+    fn read(&mut self, buff: &mut [u8]) -> Result<(), Self::Error> {
         let out = vec![0u8; buff.len()];
-        let _n = self.inner.lock().unwrap().spi_write_read(&out, buff)?;
+        let _n = self.bus.lock().unwrap().spi_write_read(&out, buff)?;
         Ok(())
     }
 }
 
-use embedded_hal::spi::blocking::{Operation, Read as _, Write as _, Transfer as _, TransferInplace};
+impl embedded_hal::spi::blocking::SpiBusFlush for SpiProxy {
+    fn flush(&mut self) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
 
-/// Default impl for transactional SPI
-impl embedded_hal::spi::blocking::Transactional<u8> for Spi {
+impl embedded_hal::spi::ErrorType for SpiProxy {
+    type Error = Error;
+}
 
-    fn exec<'a>(&mut self, operations: &mut [Operation<'a, u8>]) -> Result<(), Self::Error> {
-        for o in operations {
-            match o {
-                Operation::Write(w) => self.write(w)?,
-                Operation::Transfer(r, w) => self.transfer(r, w)?,
-                Operation::TransferInplace(b) => self.transfer_inplace(b)?,
-                Operation::Read(r) => self.read(r)?,
-            }
-        }
+impl embedded_hal::spi::Error for SpiProxy {
+    fn kind(&self) -> embedded_hal::spi::ErrorKind {
+        embedded_hal::spi::ErrorKind::Other
+    }
+}
 
-        Ok(())
+/// Spi object implements embedded-hal SPI traits for the CP2130
+pub struct Spi {
+    // TODO: use channel configuration
+    _channel: u8,
+    inner: SpiProxy,
+}
+
+impl embedded_hal::spi::blocking::SpiDevice for Spi {
+    type Bus = SpiProxy;
+
+    fn transaction<R>(
+        &mut self,
+        f: impl FnOnce(&mut Self::Bus) -> Result<R, <Self::Bus as embedded_hal::spi::ErrorType>::Error>,
+    ) -> Result<R, Self::Error> {
+        // We lock the inner mutex in every transaction, so we don't need an extra mutex here.
+        f(&mut self.inner)
```

Comment:
Just one more thing: From how the code works I gathered that this is enough to perform a safe SPI transaction for a single device, because the cp2130 will take care of handling the CS line on its own. Is that a valid assumption?

Reply:
huh, the CP2130 -can- manage the CS as part of the transaction, but, i don't typically use it that way because it doesn't allow chaining operations, which is required for the new `e-h` API... (you can enable this when [creating](https://github.com/rust-iot/rust-driver-cp2130/blob/master/src/device.rs#L450) the `Spi` instance in the current driver, but we default to [not](https://docs.rs/driver-cp2130/latest/src/driver_cp2130/device.rs.html#352-367) using it)

perhaps a better approach to match the new API would be to setup the `Spi` instance with a `CS` pin on the device that we (manually, but internally) control as part of the transaction? 

one way would be to make `Spi` generic over `CS=Auto` or `CS=Io(u8)`, then the `Auto` variant could only implement basic write/write_read, while the `Io` variant could implement the new bus API.

Reply:
(the justification here is that you need to take ownership of the bus -prior- to altering the CS pin, in which case we could technically make it generic over `CS=e_h::digital::OutputPin`)

Reply:
Ah, I see. Is that what `Spi::_channel` is supposed to handle? It has this pretty note attached:

```
// TODO: use channel configuration
```